extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpAuthenticator.java

```diff
@@ -107,28 +110,109 @@ public Uni<SecurityIdentity> apply(HttpAuthenticationMechanism mech) {
                     }
 
                 });
 
     }
 
-    private Uni<SecurityIdentity> createSecurityIdentity(RoutingContext routingContext) {
+    private Uni<SecurityIdentity> createSecurityIdentity(RoutingContext routingContext, boolean inclusiveAuthentication) {
         Uni<SecurityIdentity> result = mechanisms[0].authenticate(routingContext, identityProviderManager);
-        for (int i = 1; i < mechanisms.length; ++i) {
-            HttpAuthenticationMechanism mech = mechanisms[i];
-            result = result.onItem().transformToUni(new Function<SecurityIdentity, Uni<? extends SecurityIdentity>>() {
+
+        if (!inclusiveAuthentication) {
+            // Return the first SecurityIdentity
+            for (int i = 1; i < mechanisms.length; ++i) {
+                HttpAuthenticationMechanism mech = mechanisms[i];
+                result = result.onItem().transformToUni(new Function<SecurityIdentity, Uni<? extends SecurityIdentity>>() {
+                    @Override
+                    public Uni<SecurityIdentity> apply(SecurityIdentity identity) {
+                        if (identity != null) {
+                            return Uni.createFrom().item(identity);
+                        }
+                        return mech.authenticate(routingContext, identityProviderManager);
+                    }
+                });
+            }
+            return result;
+        } else {
+            // Have all authentication mechanisms produce SecurityIdentity
+            for (int i = 1; i < mechanisms.length; ++i) {
+                final int currentIndex = i;
+                HttpAuthenticationMechanism mech = mechanisms[i];
+                result = result.onItem().transformToUni(new Function<SecurityIdentity, Uni<? extends SecurityIdentity>>() {
+                    @Override
+                    public Uni<SecurityIdentity> apply(SecurityIdentity identity) {
+                        if (identity != null) {
+                            SecurityIdentity firstIdentity = routingContext.get(HttpSecurityUtils.SECURITY_IDENTITY_ATTRIBUTE);
+                            if (firstIdentity == null) {
+                                // If it is the very first SecurityIdentity then save it as a 'io.quarkus.security.identity' attribute on RoutingContext.
+                                firstIdentity = identity;
+                                routingContext.put(HttpSecurityUtils.SECURITY_IDENTITY_ATTRIBUTE, firstIdentity);
+                                // Get the second authentication mechanism produce SecurityIdentity
+                                return mech.authenticate(routingContext, identityProviderManager);
+                            } else {
+                                // Otherwise add it to the first SecurityIdentity's 'io.quarkus.security.identities' map attribute.
+                                // This branch will run if there are at least 3 authentication mechanism
```

Comment:
I think all this code could be simplified significantly with 'Uni.combine().all().unis(someResultsArray).combinedWith(...)'. If you are going to run all of them anyway the chaining just complicates things.